pydantic/env_settings.py

```diff
@@ -113,7 +113,10 @@ def customise_sources(
         ) -> Tuple[SettingsSourceCallable, ...]:
             return init_settings, env_settings, file_secret_settings
 
-    __config__: Config  # type: ignore
+    if TYPE_CHECKING:
+        # populated by the metaclass using the Config class defined above, annotated here to help IDEs only
+        # inside a TYPE_CHECKING block to avoid showing it in __init__ by IDEs supporting dataclass_transform
+        __config__: Type[Config]
```

Comment:
Is it intended that an annotation under `if TYPE_CHECKING` is not synthesised?  It kind of makes sense, but I'd like to make sure.

Reply:
I see that it works correctly, but I'm not completely sure that would be the right way to handle it. On runtime it's not available, so dataclasses wouldn't be able to use it, so I imagine `dataclass_transform` should also ignore it (as it currently does).

Now I'm thinking that maybe the right approach would be to use `ClassVar` instead, but then it should be updated in `BaseModel` as well, as `__config__` [and other attributes](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/main.py#L379-L395) are currently annotated there without `ClassVar` and I think those are probably class attributes, not instance attributes (not 100% sure, though).

If that's the case, I can update the annotation to use `ClassVar` and update those annotations there as well. I'll wait for more feedback on that to be certain. Maybe @samuelcolvin or @PrettyWood can confirm?

---

For other cases (not for this one) that want to declare a field that should not be included in the `__init__`, the alternative would be to declare it with a field descriptor with `Field(init=False)`, so:

```Python
__config__: Type[Config] = Field(init=False)
```

And tools supporting `dataclass_transform` (i.e. Pyright) would ignore it.

But that probably wouldn't work well for this case as the `__config__` attribute is actually populated by the metaclass, it's not really a model field.

Reply:
@tiangolo 
> And tools supporting dataclass_transform (i.e. Pyright) would ignore it.

FWIW, the PyCharm plugin ignores the field name that starts with `_`.
`dataclass_transform` should probably support an option to ignore the private attributes.

Reply:
Yep, but as attributes starting with `_` are supported by dataclasses, without additional meaning, and are included in the `__init__`, I imagine that would be the "right" behavior for them (even if that's not how it works in Pydantic).

I think the supported way for telling `dataclass_transform` to not include a field in the generated `__init__` would be with that `Field(init=False)`, that works the same way as with dataclasses. I think that would be the equivalent in this case, at least for pydantic users.

Nevertheless, this specific case is special as the `__config__` is an internal thing created by the metaclass, and it's (I think) a class attribute/variable and not really a pydantic field.

Reply:
> Yep, but as attributes starting with _ are supported by dataclasses,

Oh, I completely forgot :facepalm: 

> I think the supported way for telling dataclass_transform to not include a field in the generated __init__ would be with that Field(init=False), 

I feel that is a little hack :thinking:  (I want to hear comments by others)

>Nevertheless, this specific case is special as the __config__ is an internal thing created by the metaclass, and it's (I think) a class attribute/variable and not really a pydantic field.

I agree!! I think `ClassVar` is good. As you said it if it correct in this case.

Reply:
Using `ClassVar` looks good indeed 👍

Reply:
Great! Thanks for the feedback everyone! I updated it in `BaseSettings` and `BaseModel` to use `ClassVar` in those places.

It seems more correct indeed, and both features are preserved: correct `__init__` generated args and autocomplete/type checks for `settings.__config__.stuff`. 🎉